File: f77/build05/diffeq.f
C
      Subroutine Differential_Equations
C
C     This routine evaluates the state derivatives for a bouncing ball.
C
C     Inputs:
C       Gravity - Acceleration due to gravity [m/sec**2]
C       X               - Position of ball [m]
C       Xd              - Velocity of ball [m/sec]
C
C     Outputs:
C
C       Xdd - Acceleration of ball [m/sec**2]
C
C     Internal variables and constants:
C       Dtlmpact - Time until impact [Sec]
C
C     Declare global common
C
      Include '../driver/global.inc'
      Include '../driver/sysvars.inc'
      Equivalence (Real_Array(11), Gravity)
      Equivalence (Real_Array(14), X)
      Equivalence (Real_Array(15), Xd)
      Equivalence (Real_Array(16), Xdd)
C
C     Calculate acceleration at current time
C
      Xdd = -Gravity
C
C     Check to see if impact will occur within a maximum time step
C
      If ((X + Xd*DtMax + 0.5*Xdd*DtMax*DtMax) .le. 0.0e0) Then
C
C     If so, check to see if it is too late
C
         If ((X + Xd*DtMin + 0.5*Xdd*DtMin*DtMin) .le. 0.0e0) Then
C
C     If too late, set delay until impact to zero
C
            DtImpact = 0.0e0
C
C     Otherwise, calculate delay until actual impact
C
         Else
            DtImpact = (-Xd - Sqrt(Xd*Xd - 2.0e0*X*Xdd))/(2.0e0*X)
            If (DtImpact .lt. DtMin) Then
               DtImpact = (-Xd + Sqrt(Xd*Xd - 2.0e0*X*Xdd))/(2.0e0*X)
            End If
         End If
C
C     Schedule impact
C
         Call Schedule((Time + DtImpact), 10)
      End If
C
      Return
      End

File: gcc/build05/diffeq.c
#include "diffeq.h"
#include "../driver/global.h"
#include "../driver/sysvars.h"
#include "../driver/schedule.h"

#include <math.h>

#define gravity real_array[11].r  // Acceleration due to gravity [m/sec**2]
#define x       real_array[14].r  // Position of ball [m]
#define xd      real_array[15].r  // Velocity of ball [m/sec]
#define xdd     real_array[16].r  // Acceleration of ball [m/sec**2]

#define BOUNCE 10  // ball hits floor event

void differential_equations(void) {
  REAL dtimpact; // Time until impact [Sec]
  // Calculate acceleration at current time
  xdd = -gravity;
  // Check to see if impact will occur within a maximum time step
  if ((x + xd*dtmax + 0.5*xdd*dtmax*dtmax) <= 0.0) {
    // If so, check to see if it is too late
    if ((x+xd*dtmin + 0.5*xdd*dtmin*dtmin) <= 0.0) { // If too late, set delay until impact to zero
      dtimpact = 0.0;
    }
    else { // Otherwise, calculate delay until actual impact
      dtimpact = (-xd - sqrt(xd*xd - 2.0*x*xdd))/(2.0*x);
      if (dtimpact < dtmin)
        dtimpact = (-xd + sqrt(xd*xd-2.0*x*xdd))/(2.0*x);
    }
    // Schedule impact
    schedule((time+dtimpact),BOUNCE);
  }
}
